packages/store/src/store.ts

```diff
@@ -1,71 +1,58 @@
 import { __flush } from './scheduler'
 import { isUpdaterFunction } from './types'
-import type { AnyUpdater, Listener, Updater } from './types'
+import type { Listener } from './types'
 
-export interface StoreOptions<
-  TState,
-  TUpdater extends AnyUpdater = (cb: TState) => TState,
-> {
+export interface StoreOptions< TState> {
   /**
    * Replace the default update function with a custom one.
    */
-  updateFn?: (previous: TState) => (updater: TUpdater) => TState
+  updateFn?: (previous: TState) => (updater: (prev: TState) => TState) => TState
   /**
    * Called when a listener subscribes to the store.
    *
    * @return a function to unsubscribe the listener
    */
   onSubscribe?: (
     listener: Listener<TState>,
-    store: Store<TState, TUpdater>,
+    store: Store<TState>,
   ) => () => void
   /**
    * Called after the state has been updated, used to derive other state.
    */
   onUpdate?: () => void
 }
 
-export class Store<
-  TState,
-  TUpdater extends AnyUpdater = (cb: TState) => TState,
-> {
+export class Store<TState> {
   listeners = new Set<Listener<TState>>()
   state: TState
   prevState: TState
-  options?: StoreOptions<TState, TUpdater>
+  options?: StoreOptions<TState>
 
-  constructor(initialState: TState, options?: StoreOptions<TState, TUpdater>) {
+  constructor(initialState: TState, options?: StoreOptions<TState>) {
     this.prevState = initialState
     this.state = initialState
     this.options = options
   }
 
   subscribe = (listener: Listener<TState>) => {
     this.listeners.add(listener)
-    const unsub = this.options?.onSubscribe?.(listener, this)
+    const unsubscribe = this.options?.onSubscribe?.(listener, this)
     return () => {
       this.listeners.delete(listener)
-      unsub?.()
+      unsubscribe?.()
     }
   }
 
   /**
    * Update the store state safely with improved type checking
    */
-  setState(updater: (prevState: TState) => TState): void
-  setState(updater: TState): void
-  setState(updater: TUpdater): void
-  setState(updater: Updater<TState> | TUpdater): void {
+  setState(updater: TState | ((prevState: TState) => TState)): void {
     this.prevState = this.state
 
-    if (this.options?.updateFn) {
-      this.state = this.options.updateFn(this.prevState)(updater as TUpdater)
+    if (isUpdaterFunction(updater)) {
+      this.state = this.options?.updateFn ? this.options.updateFn(this.prevState)(updater) : updater(this.prevState)
     } else {
-      if (isUpdaterFunction(updater)) {
-        this.state = updater(this.prevState)
-      } else {
-        this.state = updater as TState
-      }
+      this.state = this.options?.updateFn ? this.options.updateFn(this.prevState)(() => updater) : updater
```

Comment:
When updater is a function, `updater: (prev: TState) => TState`

```ts
this.options.updateFn(this.prevState)(() => updater)
```
The type of () => updater is () => (prev: TState) => TState, but updateFn expects a parameter of type (prev: TState) => TState.

When updater is a value, `updater: TState`

```
ts
this.options.updateFn(this.prevState)(() => updater)
```
The type of () => updater is () => TState, but updateFn expects a parameter of type (prev: TState) => TState.

It seems like this might cause a type error. Am I missing something?

Reply:
```ts
setState(updater: TState | ((prevState: TState) => TState)): void {
  this.prevState = this.state
  
  if (this.options?.updateFn) {
    // When updateFn is defined: always convert the updater to a function.
    const functionUpdater = isUpdaterFunction(updater) 
      ? updater 
      : (prev: TState) => updater
    this.state = this.options.updateFn(this.prevState)(functionUpdater)
  } else {
    // When updateFn is not defined: handle the updater directly.
    this.state = isUpdaterFunction(updater) 
      ? updater(this.prevState) 
      : updater
  }
}
```

If I were to make changes, I think this could be a possible improvement. What do you think?

Reply:
So, we need to check both whether `updater` is a function or not **and** `updateFn` is defined or not. Either way is fine: we can check `updater`’s function type first and then `updateFn`, or vice versa.